### DISCUSSION TITLE
Schemas serialization issue while migrating to Pydantic v2 with an existing app previously using Pydantic v1

### DISCUSSION THREAD

**Discussion by tapicer** [fastapi/fastapi#10767] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()

class Address(BaseModel):
    street_name: str
    number: int
    city: str

class PartialAddress(BaseModel):
    street_name: str
    number: int

class Person(BaseModel):
    id: int
    full_name: str
    address: PartialAddress

@app.get('/test')
def test() -> Person:
    address = Address(street_name='Fake St', number=123, city='Fake City')  # in practice, this would be fetched from db
    return Person(id=1, full_name='John Doe', address=address)
```

### Description

I have a pretty big codebase using FastAPI + SQLModel and I'm migrating from Pydantic v1 to Pydantic v2 (now that SQLModel >= 0.0.14 supports it).

The `test` endpoint in the example code is a very common case of how we use schemas to return partial data from models that come from the database through SQLModel. With Pydantic v1 it's working as expected (the endpoint produces a JSON with the partial data required, `street_name` and `number` for the address), with Pydantic v2 it's throwing this exception:

`ValidationError: 1 validation error for Person address Input should be a valid dictionary or instance of PartialAddress [type=model_type, input_value=Address(street_name='Fake...r=123, city='Fake City'), input_type=Address] For further information visit https://errors.pydantic.dev/2.5/v/model_type`

I know there are some workarounds (like using Pydantic's `model_dump` to get a `dict` instead of the object), but most of them require changing several parts of the code and/or schemas, and there are still issues with several levels of models nesting. This worked as expected in Pydantic v1.

The question is: is there a simple way to make FastAPI with Pydantic v2 work the same way it did with Pydantic v1 for these cases?

### Operating System

Linux

### Operating System Details

Official Python 3.12 docker image.

### FastAPI Version

0.104.1

### Pydantic Version

2.5.2

### Python Version

3.12.0

**Comment by Ventura94** [comment]
You just have to add the optional type annotation ```Address``` and ```PartialAddress``` (```PartialAddress | Address```), in the ```address``` field of the ```Person``` schema, leaving your code as follows:

````python
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class Address(BaseModel):
    street_name: str
    number: int
    city: str


class PartialAddress(BaseModel):
    street_name: str
    number: int


class Person(BaseModel):
    id: int
    full_name: str
    address: PartialAddress | Address


@app.get('/test')
def test() -> Person:
    address = Address(street_name='Fake St', number=123, city='Fake City')  # in practice, this would be fetched from db
    return Person(id=1, full_name='John Doe', address=address)
````
I hope it is useful to you

**Reply by tapicer** [reply]
Hey @Ventura94!

Thank you for your reply. That does work as expected, however it implies I have to go through my huge codebase and add additional types here and there everywhere. I was looking for a way to make it behave like Pydantic v1 and serialize the output with the provided type (`PartialAddress` in this case).

**Reply by Ventura94** [reply, marked as answer]
Friend, the other solution I can give you is to create a base class for your schemas in the following way:
````python
class BaseSchemas(BaseModel):
    model_config = ConfigDict(
        from_attributes=True, )
````
the complete code would be this:
````python 
from fastapi import FastAPI
from pydantic import BaseModel, ConfigDict

app = FastAPI()


class BaseSchemas(BaseModel):
    model_config = ConfigDict(
        from_attributes=True, )


class Address(BaseSchemas):
    street_name: str
    number: int
    city: str


class PartialAddress(BaseSchemas):
    street_name: str
    number: int


class Person(BaseSchemas):
    id: int
    full_name: str
    address: PartialAddress


@app.get('/test')
def test() -> Person:
    address = Address(street_name='Fake St', number=123, city='Fake City')  # in practice, this would be fetched from db
    return Person(id=1, full_name='John Doe', address=address)
````
Obviously it requires a change, but from my point of view it is required and I think this would be the simplest way.
